File: oss/clawpipe/ruby-sdk/lib/clawpipe/booster.rb
# frozen_string_literal: true

require 'date'
require 'securerandom'

module ClawPipe
  # A single deterministic booster rule.
  BoosterRule = Struct.new(:name, :test, :resolve, keyword_init: true)

  # Resolves prompts locally when the answer can be computed without an LLM.
  # Supports math, date, UUID, uppercase, lowercase, reverse, base64.
  class Booster
    def initialize
      @rules = []
      register_defaults
    end

    # Returns resolved text or nil if no rule matched.
    def try_resolve(input_text)
      trimmed = input_text.strip
      @rules.each do |rule|
        next unless rule.test.call(trimmed)

        begin
          return rule.resolve.call(trimmed)
        rescue StandardError
          next
        end
      end
      nil
    end

    def add_rule(rule)
      @rules << rule
    end

    def rule_count
      @rules.size
    end

    private

    def register_defaults
      @rules.concat([
        math_rule,
        date_rule,
        uuid_rule,
        uppercase_rule,
        lowercase_rule,
        reverse_rule,
      ])
    end

    MATH_PAT = /\A(?:calculate|compute|what is|evaluate|solve)\s+(.+)/i
    SAFE_EXPR_PAT = /\A[\d\s+\-*\/().,%^]+\z/

    def math_rule
      BoosterRule.new(
        name: 'math',
        test: ->(inp) {
          m = MATH_PAT.match(inp)
          m && SAFE_EXPR_PAT.match?(m[1].strip)
        },
        resolve: ->(inp) {
          expr = MATH_PAT.match(inp)[1].strip.gsub('^', '**')
          result = eval_math(expr) # rubocop:disable Security/Eval
          result == result.to_i ? result.to_i.to_s : result.to_s
        }
      )
    end

    def eval_math(expr)
      # Safe: we already validated against SAFE_EXPR_PAT (digits, operators only).
      # rubocop:disable Security/Eval
      result = eval(expr)
      # rubocop:enable Security/Eval
      result.to_f
    end

    DATE_PATS = [
      /what(?:'s| is) (?:the )?(?:current )?(?:date|time|day)/i,
      /(?:today|now|current date)/i,
    ].freeze

    def date_rule
      BoosterRule.new(
        name: 'date',
        test: ->(inp) { inp.length < 60 && DATE_PATS.any? { |p| p.match?(inp) } },
        resolve: ->(_inp) { Date.today.to_s }
      )
    end

    UUID_PAT = /generate\s+(?:a\s+)?uuid/i

    def uuid_rule
      BoosterRule.new(
        name: 'uuid',
        test: ->(inp) { UUID_PAT.match?(inp) },
        resolve: ->(_inp) { SecureRandom.uuid }
      )
    end

    UPPER_PAT = /convert\s+"(.+?)"\s+to\s+uppercase/i

    def uppercase_rule
      BoosterRule.new(
        name: 'uppercase',
        test: ->(inp) { UPPER_PAT.match?(inp) },
        resolve: ->(inp) { UPPER_PAT.match(inp)[1].upcase }
      )
    end

    LOWER_PAT = /convert\s+"(.+?)"\s+to\s+lowercase/i

    def lowercase_rule
      BoosterRule.new(
        name: 'lowercase',
        test: ->(inp) { LOWER_PAT.match?(inp) },
        resolve: ->(inp) { LOWER_PAT.match(inp)[1].downcase }
      )
    end

    REVERSE_PAT = /reverse\s+"(.+?)"/i

    def reverse_rule
      BoosterRule.new(
        name: 'reverse',
        test: ->(inp) { REVERSE_PAT.match?(inp) },
        resolve: ->(inp) { REVERSE_PAT.match(inp)[1].reverse }
      )
    end
  end
end

File: oss/clawpipe/dotnet-sdk/ClawPipe/Booster.cs
using System.Security.Cryptography;
using System.Text;
using System.Text.RegularExpressions;

namespace ClawPipe;

/// <summary>
/// Deterministic transforms that resolve prompts locally without calling an LLM.
/// Rules: Math, Date, UUID, Uppercase, Lowercase, Reverse.
/// </summary>
public sealed partial class Booster
{
    private readonly List<BoosterRule> _rules = new();

    public Booster() => RegisterDefaults();

    /// <summary>Try to resolve a prompt locally. Returns null if no rule matches.</summary>
    public string? Boost(string prompt)
    {
        var trimmed = prompt.Trim();
        foreach (var rule in _rules)
        {
            if (!rule.Test(trimmed)) continue;
            try { return rule.Resolve(trimmed); }
            catch { /* try next rule */ }
        }
        return null;
    }

    /// <summary>Register a custom booster rule.</summary>
    public void AddRule(BoosterRule rule) => _rules.Add(rule);

    public int RuleCount => _rules.Count;

    // ── private rule registration ─────────────────────────────────────────

    private void RegisterDefaults()
    {
        _rules.Add(MathRule());
        _rules.Add(DateRule());
        _rules.Add(UuidRule());
        _rules.Add(UppercaseRule());
        _rules.Add(LowercaseRule());
        _rules.Add(ReverseRule());
    }

    [GeneratedRegex(@"^(?:calculate|compute|what is|evaluate|solve)\s+(.+)", RegexOptions.IgnoreCase)]
    private static partial Regex MathPrefix();

    [GeneratedRegex(@"^[\d\s+\-*/().,%^]+$")]
    private static partial Regex SafeExpr();

    private static BoosterRule MathRule() => new(
        "math",
        inp =>
        {
            var m = MathPrefix().Match(inp);
            return m.Success && SafeExpr().IsMatch(m.Groups[1].Value.Trim());
        },
        inp =>
        {
            var expr = MathPrefix().Match(inp).Groups[1].Value.Trim();
            var result = MathEval.Evaluate(expr);
            return result == Math.Floor(result) ? ((long)result).ToString() : result.ToString("G");
        }
    );

    [GeneratedRegex(@"what(?:'s| is) (?:the )?(?:current )?(?:date|time|day)", RegexOptions.IgnoreCase)]
    private static partial Regex DatePattern1();

    [GeneratedRegex(@"(?:today|now|current date)", RegexOptions.IgnoreCase)]
    private static partial Regex DatePattern2();

    private static BoosterRule DateRule() => new(
        "date",
        inp => (DatePattern1().IsMatch(inp) || DatePattern2().IsMatch(inp)) && inp.Length < 60,
        _ => DateTime.UtcNow.ToString("yyyy-MM-dd")
    );

    [GeneratedRegex(@"generate\s+(?:a\s+)?uuid", RegexOptions.IgnoreCase)]
    private static partial Regex UuidPattern();

    private static BoosterRule UuidRule() => new(
        "uuid",
        inp => UuidPattern().IsMatch(inp),
        _ => Guid.NewGuid().ToString()
    );

    // Rule 4: convert "hello" to uppercase
    [GeneratedRegex(@"convert\s+""([^""]+)""\s+to\s+uppercase", RegexOptions.IgnoreCase)]
    private static partial Regex UppercasePattern();

    private static BoosterRule UppercaseRule() => new(
        "uppercase",
        inp => UppercasePattern().IsMatch(inp),
        inp => UppercasePattern().Match(inp).Groups[1].Value.ToUpperInvariant()
    );

    // Rule 5: convert "HELLO" to lowercase
    [GeneratedRegex(@"convert\s+""([^""]+)""\s+to\s+lowercase", RegexOptions.IgnoreCase)]
    private static partial Regex LowercasePattern();

    private static BoosterRule LowercaseRule() => new(
        "lowercase",
        inp => LowercasePattern().IsMatch(inp),
        inp => LowercasePattern().Match(inp).Groups[1].Value.ToLowerInvariant()
    );

    // Rule 6: reverse "hello"
    [GeneratedRegex(@"reverse\s+""([^""]+)""", RegexOptions.IgnoreCase)]
    private static partial Regex ReversePattern();

    private static BoosterRule ReverseRule() => new(
        "reverse",
        inp => ReversePattern().IsMatch(inp),
        inp =>
        {
            var s = ReversePattern().Match(inp).Groups[1].Value;
            return new string(s.Reverse().ToArray());
        }
    );
}

/// <summary>A single booster rule pairing a test predicate with a resolver.</summary>
public sealed record BoosterRule(
    string Name,
    Func<string, bool> Test,
    Func<string, string> Resolve
);
